### PR TITLE
Update rpc client header

### PR DESCRIPTION
**Pull request by samricotta**
Moving the changes from issue[ #7270](https://github.com/tendermint/tendermint/pull/7270) into main. This introduces header and `header_by_hash queries` to the RPC client. This is useful for Evmos JSON-RPC to avoid querying the block endpoint for the `eth_header` and `eth_header_by_hash` JSON-RPC endpoints

Note: This is part of the ABCI++ Prepare/Process Proposal Release Plan https://github.com/tendermint/tendermint/issues/9091

### CODE REVIEW

File: CHANGELOG_PENDING.md [hunk 1]
```diff
@@ -29,6 +29,8 @@ Friendly reminder, we have a [bug bounty program](https://hackerone.com/tendermi
 
 ### IMPROVEMENTS
 
+[abci/rpc] \#9276 `header` and `header_by_hash` queries to the RPC client
```
Comment (cmwaters) [inline CHANGELOG_PENDING.md:32]:
```suggestion
[rpc] \#9276 Add `header` and `header_by_hash` queries to the RPC layer
```

File: rpc/openapi/openapi.yaml [hunk 2]
```diff
@@ -637,6 +637,64 @@ paths:
...
+        - Info
+      description: |
+        Get Header By Hash.
+      responses:
+        "200":
+          description: Header informations.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/HeaderResponse"
```
Comment (cmwaters) [inline rpc/openapi/openapi.yaml:691]:
```suggestion
                $ref: "#/components/schemas/BlockHeader"
```

File: rpc/openapi/openapi.yaml [hunk 3]
```diff
@@ -637,6 +637,64 @@ paths:
...
+        - Info
+      description: |
+        Get Header.
+      responses:
+        "200":
+          description: Header informations.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/HeaderResponse"
```
Comment (cmwaters) [inline rpc/openapi/openapi.yaml:662]:
```suggestion
                $ref: "#/components/schemas/BlockHeader"
```

File: store/store.go [hunk 4]
```diff
@@ -196,6 +196,26 @@ func (bs *BlockStore) LoadBlockMeta(height int64) *types.BlockMeta {
 	return blockMeta
 }
 
+// LoadBlockMetaByHash returns the blockmeta who's header corresponds to the given
+// hash. If none is found, returns nil.
+func (bs *BlockStore) LoadBlockMetaByHash(hash []byte) *types.BlockMeta {
```
Comment (cmwaters) [inline store/store.go:201]: Want to add a test for this as well? (or add it in an existing test?)

**Review by williambanfield — COMMENTED** [review]
Looks good, would echo Callum's request for a test of `LoadBlockMetaByHash`

File: CHANGELOG_PENDING.md [hunk 5]
```diff
@@ -36,8 +36,11 @@ Friendly reminder, we have a [bug bounty program](https://hackerone.com/tendermi
 
 ### IMPROVEMENTS
 
+- [rpc] \#9276 `header` and `header_by_hash` queries to the RPC client
 - [abci] \#5706 Added `AbciVersion` to `RequestInfo` allowing applications to check ABCI version when connecting to Tendermint. (@marbar3778)
 
+[rpc] \#9276 `header` and `header_by_hash` queries to the RPC client
+
```
Comment (sergio-mena) [inline CHANGELOG_PENDING.md:43]: Is this supposed to be here?

File: CHANGELOG_PENDING.md [hunk 6]
```diff
@@ -36,8 +36,11 @@ Friendly reminder, we have a [bug bounty program](https://hackerone.com/tendermi
 
 ### IMPROVEMENTS
 
+- [rpc] \#9276 `header` and `header_by_hash` queries to the RPC client
```
Comment (sergio-mena) [inline CHANGELOG_PENDING.md:39]:
```suggestion
- [rpc] \#9276 Added `header` and `header_by_hash` queries to the RPC client
```

File: rpc/core/blocks.go [hunk 7]
```diff
@@ -75,6 +76,38 @@ func filterMinMax(base, height, min, max, limit int64) (int64, int64, error) {
 	return min, max, nil
 }
 
+// Header gets block header at a given height.
+// If no height is provided, it will fetch the latest header.
+// More: https://docs.tendermint.com/master/rpc/#/Info/header
```
Comment (sergio-mena) [inline rpc/core/blocks.go:81]: The link contains `master`, should it be renamed to `main`?

File: store/store_test.go [hunk 8]
```diff
@@ -552,6 +553,28 @@ func TestLoadBlockMeta(t *testing.T) {
...
+	block = state.MakeBlock(state.LastBlockHeight+1, test.MakeNTxs(state.LastBlockHeight+1, 10), new(types.Commit), nil, state.Validators.GetProposer().Address)
+	partSet, err := block.MakePartSet(2)
+	require.NoError(t, err)
+	seenCommit := makeTestCommit(1, tmtime.Now())
+	bs.SaveBlock(block, partSet, seenCommit)
+
+	baseBlock := bs.LoadBlockMetaByHash(block.Hash())
+	assert.EqualValues(t, block.Header.Height, baseBlock.Header.Height)
+	assert.EqualValues(t, block.Header.LastBlockID, baseBlock.Header.LastBlockID)
+	assert.EqualValues(t, block.Header.ChainID, baseBlock.Header.ChainID)
```
Comment (sergio-mena) [inline store/store_test.go:575]: Can you flip one bit of the hash, repeat the call, and check we now get `nil`?

**Review by sergio-mena — APPROVED** [review]
LGTM

**Pull request merged**